notifications_utils/recipient_validation/errors.py

```diff
@@ -1,7 +1,15 @@
+from enum import StrEnum, auto
+
+
 class InvalidRecipientError(Exception):
     message = "Not a valid recipient address"
+    code = "UNKNOWN"
 
-    def __init__(self, message=None):
+    def __init__(self, message: str = None, code: str = None):
+        if code:
+            self.code = code
+            if not message:
+                message = code
         super().__init__(message or self.message)
```

Comment:
I had a hard time reasoning about this conditional.

I think this is equivalent (and makes the precedence clearer) but I still can’t see where `"UNKNOWN"` or `self.code` is ever used – do we need those lines?
```suggestion
        message = message or code or self.message
        self.code = code or "UNKNOWN"
        super().__init__(message)
```

Reply:
that's fair about the conditional.

i guess i was wondering about making this the same across all the three email/phone/address validation errors but that didn't happen - i can probs move `self.code` to just be in `InvalidPhoneError` for now. it's hard to imagine us handling `InvalidRecipientError` where we both:

* care about specific error codes (eg too long, too short, invalid country code) more than just "what is the error message to display"
* don't know whether we're validating an sms, email, or letter

i'll update the base class to remove `self.code` completely and just keep the default message there for now

Reply:
Done (squashed into https://github.com/alphagov/notifications-utils/pull/1125/commits/a0ebe1a9ee08b97b23315acb082f4c8e293bc5f1)